src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -338,141 +272,123 @@ internal void MeasurementsCompleted(Instrument instrument, object state)
-            if (metrics == null)
-            {
-                // TODO: log
-                return;
-            }
 
-            if (metrics.Count == 1)
+            if (this.compositeMetricReader == null)
             {
-                // special casing the common path
-                // as this is faster than the
-                // foreach, when count is 1.
-                metrics[0].UpdateDouble(value, tagsRos);
+                if (state is not List<Metric> metrics)
+                {
+                    // TODO: log
+                    return;
+                }
+
+                this.reader.RecordDoubleMeasurement(metrics, value, tagsRos);
             }
             else
             {
-                foreach (var metric in metrics)
+                if (state is not List<List<Metric>> metricsSuperList)
                 {
-                    metric.UpdateDouble(value, tagsRos);
+                    // TODO: log
+                    return;
                 }
+
+                this.compositeMetricReader.RecordDoubleMeasurements(metricsSuperList, value, tagsRos);
             }
         }
 
         internal void MeasurementRecordedLong(Instrument instrument, long value, ReadOnlySpan<KeyValuePair<string, object>> tagsRos, object state)
         {
-            // Get Instrument State
-            var metrics = state as List<Metric>;
-
             Debug.Assert(instrument != null, "instrument must be non-null.");
-            if (metrics == null)
-            {
-                // TODO: log
-                return;
-            }
 
-            if (metrics.Count == 1)
+            if (this.compositeMetricReader == null)
             {
-                // special casing the common path
-                // as this is faster than the
-                // foreach, when count is 1.
-                metrics[0].UpdateLong(value, tagsRos);
+                if (state is not List<Metric> metrics)
+                {
+                    // TODO: log
+                    return;
+                }
+
+                this.reader.RecordLongMeasurement(metrics, value, tagsRos);
             }
             else
             {
-                foreach (var metric in metrics)
+                if (state is not List<List<Metric>> metricsSuperList)
                 {
-                    metric.UpdateLong(value, tagsRos);
+                    // TODO: log
+                    return;
                 }
+
+                this.compositeMetricReader.RecordLongMeasurements(metricsSuperList, value, tagsRos);
             }
         }
 
         internal void MeasurementRecordedLongSingleStream(Instrument instrument, long value, ReadOnlySpan<KeyValuePair<string, object>> tagsRos, object state)
         {
-            // Get Instrument State
-            var metric = state as Metric;
-
             Debug.Assert(instrument != null, "instrument must be non-null.");
-            if (metric == null)
+
+            if (this.compositeMetricReader == null)
```

Comment:
the name of the method no longer makes sense as it handles single stream(Metric) and multiple (List<metric>).
We should make 3 separate methods for Metric, List of Metic, List of (list of metric). Might save all the if-checks in hot path.
Good for a follow up to avoid making this PR bigger.